SleepEngine/Engine/Components/HierarchyNode.cpp

```diff
@@ -7,57 +7,75 @@ BEGIN_NAMESPACE_SLEEP
 
 HierarchyNode::HierarchyNode()
     : Base(Component::getComponentTypeID<HierarchyNode>())
+    , m_parent(nullptr)
 {}
 
 Object* HierarchyNode::addChild(std::unique_ptr<Object>&& child)
 {
-    auto* childNode = child->getComponent<HierarchyNode>();
+    auto* const childHandle = child.get();
 
-    if(!childNode)
-    {
-        childNode = child->addComponent<HierarchyNode>();
-    }
-
-    if (childNode->getParent())
-    {
-        childNode->detachFromParent();
-    }
-    childNode->setParent(m_object);
+    setUpChild(*childHandle);
 
-    auto* childHandle = child.get();
     m_children.emplace_back(std::move(child));
     return childHandle;
 }
 
 void HierarchyNode::removeChild(Object* child)
 {
-    auto isChildToDelete = [child] (std::unique_ptr<Object> const& element)
+    auto const isChildToDelete = [child] (std::unique_ptr<Object> const& element)
     {
         return element.get() == child;
     };
-    LOG_AND_ASSERT_ERROR(containsIf(m_children, isChildToDelete), "not my child!");
-    removeIf(m_children, isChildToDelete);
+
+    auto const childToDelete = findIf(m_children, isChildToDelete);
+
+    if (childToDelete != m_children.cend())
+    {
+        LOG_AND_FAIL("'removeChild' called for object that is not out child");
+        return;
+    }
+
+    m_children.erase(childToDelete);
 }
 
 void HierarchyNode::detachFromParent()
 {
-    if (!m_object)
+    if (!m_parent)
     {
         LOG_AND_FAIL("parent is null");
         return;
     }
 
-    auto* parentNode = m_object->getComponent<HierarchyNode>();
-    parentNode->removeChild(m_object);
-    m_object = nullptr;
+    auto* const parentNode = m_parent->getComponent<HierarchyNode>();
+    parentNode->removeChild(m_parent);
+    m_parent = nullptr;
 }
 
 void HierarchyNode::update(float dt)
 {
     Base::update(dt);
-    for(auto& childNodeObject : m_children)
+
+    for (auto& childNodeObject : m_children)
     {
         childNodeObject->update(dt);
     }
 }
+
+void HierarchyNode::setUpChild(Object& child) const

```

Comment:
this function is used only once, and has a poor name - too generic ,and i presume - purpose. You've basically just moved all code from addChild to here. It is also used only once. I think it hinders readability and elegance, i'd suggest you just remove it. As of now

Reply:
Ok, done